atlasdb-tests-shared/src/main/java/com/palantir/atlasdb/timestamp/TimestampServiceTests.java

```diff
@@ -0,0 +1,141 @@
+
+        long firstUpperBound = timestampRanges.get(0).getUpperBound();
+        long secondLowerBound = timestampRanges.get(1).getLowerBound();
+
+        MatcherAssert.assertThat(firstUpperBound, is(lessThan(secondLowerBound)));
+    }
+
+    public static void canRequestMoreTimestampsThanAreAllocatedAtOnce(TimestampService timestampService) {
+        for (int i = 0; i < ONE_MILLION / 1000; i++) {
+            timestampService.getFreshTimestamps(1000);
+        }
+
+        MatcherAssert.assertThat(timestampService.getFreshTimestamp(), is(ONE_MILLION + 1));
+    }
+
+    public static void willNotHandOutTimestampsEarlierThanAFastForward(
+            TimestampMigrationService timestampMigrationService, TimestampService timestampService) {
+        timestampMigrationService.fastForwardTimestamp(TWO_MILLION);
+
+        MatcherAssert.assertThat(
+                timestampService.getFreshTimestamp(),
+                is(greaterThan(TWO_MILLION)));
+    }
+
+    public static void willDoNothingWhenFastForwardToEarlierTimestamp(
+            TimestampMigrationService timestampMigrationService,
+            TimestampService timestampService) {
+        timestampMigrationService.fastForwardTimestamp(TWO_MILLION);
+        long ts1 = timestampService.getFreshTimestamp();
+        timestampMigrationService.fastForwardTimestamp(ONE_MILLION);
+        long ts2 = timestampService.getFreshTimestamp();
+        MatcherAssert.assertThat(ts2, greaterThan(TWO_MILLION));
+        MatcherAssert.assertThat(ts2, greaterThan(ts1));
+    }
+
+    public static void canReturnManyUniqueTimestampsInParallel(TimestampService timestampService)
+            throws InterruptedException, TimeoutException {
+        Set<Long> uniqueTimestamps = new ConcurrentSkipListSet<>();
+
+        repeat(TWO_MILLION, () -> uniqueTimestamps.add(timestampService.getFreshTimestamp()));
+
+        MatcherAssert.assertThat(uniqueTimestamps.size(), is((int) TWO_MILLION));
+    }
+
+    private static void repeat(long count, Runnable task) throws InterruptedException, TimeoutException {
+        for (int i = 0; i < count; i++) {
+            executor.submit(task);
+        }
+
+        executor.shutdown();
+        executor.awaitTermination(1, TimeUnit.MINUTES);
+
+        if (!executor.isTerminated()) {
+            throw new TimeoutException("Timed out waiting for the executor to terminate");
+        }
+    }
+
+    public static void shouldThrowIfRequestingNegativeNumbersOfTimestamps(TimestampService timestampService) {
+        Assertions.assertThatThrownBy(() -> timestampService.getFreshTimestamps(-1))
+                .isInstanceOf(IllegalArgumentException.class);
+    }
+
+    public static void shouldThrowIfRequestingZeroTimestamps(TimestampService timestampService) {
+        Assertions.assertThatThrownBy(() -> timestampService.getFreshTimestamps(0))
+                .isInstanceOf(IllegalArgumentException.class);
+    }
```

Comment:
Thanks for adding the last 2 tests.

At some point we should migrate all the Hamcrest stuff here to AssertJ, though I realise that might not be a high priority at the moment. (The fact that it's easy to identify that someone has added these separately from the rest of the tests, even with no knowledge of how the tests looked like previously is probably some kind of a flag.)